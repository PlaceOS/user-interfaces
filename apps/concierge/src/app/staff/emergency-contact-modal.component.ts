import { Component, Inject, ViewChild } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
import { EmergencyContact } from './emergency-contacts.component';
import { FormControl, FormGroup } from '@angular/forms';
import { OrganisationService } from '@placeos/organisation';
import { showMetadata, updateMetadata } from '@placeos/ts-client';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { filter, map, shareReplay, switchMap, take } from 'rxjs/operators';
import { notifySuccess, randomString } from '@placeos/common';
import { CustomTooltipComponent } from '@placeos/components';

@Component({
    selector: 'emergency-contact-modal',
    template: `
        <header>
            <h2>{{ contact ? 'Edit' : 'New' }} Emergency Contact</h2>
            <div class="flex-1 w-0"></div>
            <button icon mat-dialog-close *ngIf="!loading">
                <app-icon>close</app-icon>
            </button>
        </header>
        <main class="p-4 w-[36rem]" *ngIf="!loading; else load_state">
            <form [formGroup]="form" class="space-y-4">
                <a-user-search-field
                    ngModel
                    (ngModelChange)="setUser($event)"
                    [ngModelOptions]="{ standalone: true }"
                    class="mb-4"
                ></a-user-search-field>
                <div class="flex flex-col">
                    <label for="name">Name:</label>
                    <mat-form-field appearance="outline">
                        <input
                            matInput
                            formControlName="name"
                            placeholder="Full name"
                        />
                    </mat-form-field>
                </div>
                <div class="flex items-center space-x-2">
                    <div class="flex flex-col flex-1">
                        <label for="email">Email:</label>
                        <mat-form-field appearance="outline">
                            <input
                                matInput
                                formControlName="email"
                                type="email"
                                placeholder="Email address"
                            />
                        </mat-form-field>
                    </div>
                    <div class="flex flex-col flex-1">
                        <label for="email">Phone:</label>
                        <mat-form-field appearance="outline">
                            <input
                                matInput
                                formControlName="phone"
                                type="tel"
                                placeholder="Emergency contact number"
                            />
                        </mat-form-field>
                    </div>
                </div>
                <div class="flex flex-col">
                    <label for="name">Level:</label>
                    <mat-form-field appearance="outline">
                        <mat-select
                            formControlName="zone"
                            placeholder="All Levels"
                        >
                            <mat-option value="">All Levels</mat-option>
                            <mat-option
                                *ngFor="let level of levels | async"
                                [value]="level.id"
                            >
                                {{ level.display_name || level.name }}
                            </mat-option>
                        </mat-select>
                    </mat-form-field>
                </div>
                <div class="flex flex-col">
                    <label for="email">Roles:</label>
                    <div class="flex items-center space-x-2">
                        <mat-form-field
                            class="no-subscript flex-1"
                            appearance="outline"
                        >
                            <mat-select
                                multiple
                                formControlName="roles"
                                placeholder="Select roles"
                            >
                                <ng-container
                                    *ngFor="
                                        let role of (data | async)?.roles || []
                                    "
                                >
                                    <mat-option *ngIf="role" [value]="role">
                                        {{ role }}
                                    </mat-option>
                                </ng-container>
                            </mat-select>
                        </mat-form-field>
                        <button
                            btn
                            matRipple
                            class="space-x-2"
                            customTooltip
                            [content]="role_form"
                        >
                            <app-icon>add</app-icon>
                            <div class="pr-2">Add New Role</div>
                        </button>
                    </div>
                </div>
            </form>
        </main>
        <footer
            *ngIf="!loading"
            class="flex justify-center items-center p-2 border-t border-base-200"
        >
            <button btn matRipple class="w-32" (click)="save()">Save</button>
        </footer>
        <ng-template #load_state>
            <main
                loading
                class="h-64 flex flex-col items-center justify-center"
            >
                <mat-spinner [diameter]="48" class="mb-4"></mat-spinner>
                <p>Saving contact details...</p>
            </main>
        </ng-template>
        <ng-template #role_form>
            <div class="bg-base-100 p-4 rounded ">
                <mat-form-field appearance="outline">
                    <input
                        matInput
                        [(ngModel)]="role_name"
                        placeholder="Role name"
                    />
                </mat-form-field>
                <button btn matRipple class="w-full" (click)="addRole()">
                    Save Role
                </button>
            </div>
        </ng-template>
    `,
    styles: [``],
})
export class EmergencyContactModalComponent {
    private _changes = new BehaviorSubject(0);

    public loading = false;
    public role_name: string;
    public readonly contact?: EmergencyContact = this._data;
    public readonly data = combineLatest([
        this._org.active_building,
        this._changes,
    ]).pipe(
        filter(([bld]) => !!bld),
        switchMap(([bld]) => showMetadata(bld.id, 'emergency_contacts')),
        map(({ details }) => (details as any) || { roles: [], contacts: [] }),
        shareReplay(1)
    );
    public readonly form = new FormGroup({
        id: new FormControl(this._data?.id || `ecntct-${randomString(8)}`),
        name: new FormControl(this._data?.name || ''),
        email: new FormControl(this._data?.email || ''),
        phone: new FormControl(this._data?.phone || ''),
        zone: new FormControl(this._data?.zone || ''),
        roles: new FormControl(this._data?.roles || []),
    });
    /** List of levels for the active building */
    public readonly levels = this._org.active_levels;

    @ViewChild(CustomTooltipComponent) private _tooltip: CustomTooltipComponent;

    constructor(
        @Inject(MAT_DIALOG_DATA) private _data: EmergencyContact | undefined,
        private _dialog_ref: MatDialogRef<EmergencyContactModalComponent>,
        private _org: OrganisationService
    ) {}

    public async addRole() {
        if (!this.role_name) return;
        this._tooltip.close();
        this.loading = true;
        this._dialog_ref.disableClose = true;
        const data: any = await this.data.pipe(take(1)).toPromise();
        await updateMetadata(this._org.building.id, {
            name: 'emergency_contacts',
            description: 'Emergency Contacts',
            details: {
                roles: [...(data.roles || []), this.role_name].filter(
                    (_) => !!_
                ),
                contacts: data.contacts,
            },
        }).toPromise();
        this._changes.next(0);
        this.form.patchValue({
            roles: [...(this.form.value.roles || []), this.role_name],
        });
        this.role_name = '';
        this.loading = false;
        this._dialog_ref.disableClose = false;
    }

    public setUser(user: any) {
        this.form.patchValue({
            name: user?.name,
            email: user?.email,
            phone: user?.phone,
        });
    }

    public async save() {
        this.loading = true;
        this._dialog_ref.disableClose = true;
        const data: any = await this.data.pipe(take(1)).toPromise();
        const contacts = data?.contacts || [];
        const new_contacts = [
            ...contacts.filter((_) => _.id !== this.contact?.id),
            this.form.value,
        ].sort((a, b) => a.name.localeCompare(b.name));
        await updateMetadata(this._org.building.id, {
            name: 'emergency_contacts',
            description: 'Emergency Contacts',
            details: { roles: data.roles || [], contacts: new_contacts },
        }).toPromise();
        this._dialog_ref.disableClose = true;
        notifySuccess('Successfully updated emergency contacts.');
        this.loading = false;
        this._dialog_ref.close();
    }
}
