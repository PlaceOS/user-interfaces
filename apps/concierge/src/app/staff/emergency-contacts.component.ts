import { Component } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { OrganisationService } from '@placeos/organisation';
import { showMetadata, updateMetadata } from '@placeos/ts-client';
import { BehaviorSubject, combineLatest, of } from 'rxjs';
import { filter, map, shareReplay, switchMap, take } from 'rxjs/operators';
import { EmergencyContactModalComponent } from './emergency-contact-modal.component';
import { notify, notifySuccess, openConfirmModal } from '@placeos/common';
import { RoleManagementModalComponent } from './role-management-modal.component';
import { Clipboard } from '@angular/cdk/clipboard';

export interface EmergencyContact {
    id: string;
    email: string;
    name: string;
    phone: string;
    roles: string[];
    zone: string;
}

export interface EmergencyContactData {
    contacts: EmergencyContact[];
    roles: string[];
}

@Component({
    selector: '[app-emergency-contacts]',
    template: `
        <app-topbar></app-topbar>
        <div class="flex flex-1 h-px">
            <app-sidebar></app-sidebar>
            <main class="flex flex-col flex-1 w-1/2 h-full">
                <section topbar class="px-8 py-4 flex flex-col">
                    <div class="flex items-center justify-between">
                        <h2 class="text-2xl font-medium">Emergency Contacts</h2>
                        <div class="flex items-center space-x-2">
                            <mat-form-field
                                class="no-subscript"
                                appearance="outline"
                            >
                                <app-icon class="text-2xl" matPrefix>
                                    search
                                </app-icon>
                                <input
                                    matInput
                                    [(ngModel)]="search"
                                    placeholder="Filter contacts..."
                                />
                            </mat-form-field>
                            <button
                                btn
                                matRipple
                                class="space-x-2"
                                (click)="editContact()"
                            >
                                <app-icon class="text-2xl">add</app-icon>
                                <div class="pr-2">Add Contact</div>
                            </button>
                        </div>
                    </div>
                    <div class="flex items-center justify-between py-2 mt-2">
                        <mat-form-field
                            class="no-subscript"
                            appearance="outline"
                        >
                            <mat-select
                                [ngModel]="role_filter.getValue()"
                                (ngModelChange)="role_filter.next($event)"
                                placeholder="All Roles"
                            >
                                <mat-option value="">All Roles</mat-option>
                                <mat-option
                                    *ngFor="let role of (roles | async) || []"
                                    [value]="role"
                                >
                                    {{ role }}
                                </mat-option>
                            </mat-select>
                        </mat-form-field>
                        <div class="flex items-center space-x-2">
                            <button
                                icon
                                matRipple
                                class="h-12 w-12 bg-secondary text-secondary-content rounded"
                                matTooltip="Manage Roles"
                                (click)="manageRoles()"
                            >
                                <app-icon>list_alt</app-icon>
                            </button>
                        </div>
                    </div>
                </section>
                <section class="w-full h-1/2 flex-1 overflow-auto px-8">
                    <simple-table
                        class="min-w-[52rem] block text-sm"
                        [data]="filtered_contacts"
                        [filter]="search"
                        [empty_message]="
                            search
                                ? 'No matching contacts'
                                : 'No emergency contacts for this building'
                        "
                        [columns]="[
                            {
                                key: 'name',
                                name: 'Person',
                                content: person_template
                            },
                            {
                                key: 'roles',
                                name: 'Roles',
                                content: roles_template,
                                sortable: false
                            },
                            {
                                key: 'zone',
                                name: 'Zone',
                                content: zone_template,
                                size: '8rem',
                                sortable: false
                            },
                            {
                                key: 'actions',
                                name: ' ',
                                content: actions_template,
                                size: '6rem',
                                sortable: false
                            }
                        ]"
                        [sortable]="true"
                    ></simple-table>
                    <div class="w-full h-12"></div>
                    <ng-template #person_template let-row="row">
                        <button
                            class="px-4 py-2 text-left leading-tight"
                            (click)="copyToClipboard(row.email)"
                        >
                            <div class="">{{ row.name }}</div>
                            <div class="text-[0.625rem] opacity-30 font-mono">
                                {{ row.email }}
                            </div>
                        </button>
                    </ng-template>
                    <ng-template #roles_template let-data="data">
                        <div class="flex flex-wrap p-2">
                            <span
                                class="m-1 py-1 px-2 rounded-2xl text-xs font-mono bg-info text-info-content"
                                *ngFor="let role of data"
                            >
                                {{ role }}
                            </span>
                        </div>
                    </ng-template>
                    <ng-template #zone_template let-data="data">
                        <div class="p-4">
                            {{ data ? (data | level)?.display_name : 'All' }}
                        </div>
                    </ng-template>
                    <ng-template #actions_template let-row="row">
                        <div
                            class="flex items-center justify-end w-full space-x-2 p-2"
                        >
                            <button
                                icon
                                matRipple
                                matTooltip="Edit Emergency Contact"
                                (click)="editContact(row)"
                            >
                                <app-icon>edit</app-icon>
                            </button>
                            <button
                                icon
                                matRipple
                                class="text-error"
                                (click)="removeContact(row)"
                                matTooltip="Remove Emergency Contact"
                            >
                                <app-icon>delete</app-icon>
                            </button>
                        </div>
                    </ng-template>
                </section>
            </main>
        </div>
    `,
    styles: [
        `
            :host {
                display: flex;
                flex-direction: column;
                height: 100%;
                width: 100%;
                background-color: var(--b1);
            }
        `,
    ],
})
export class EmergencyContactsComponent {
    private _change = new BehaviorSubject<number>(0);

    public search = '';
    public readonly role_filter = new BehaviorSubject<string>('');
    public readonly data = combineLatest([
        this._org.active_building,
        this._change,
    ]).pipe(
        filter(([bld]) => !!bld),
        switchMap(([bld]) => showMetadata(bld.id, 'emergency_contacts')),
        map(({ details }) => (details as any) || { roles: [], contacts: [] }),
        shareReplay(1)
    );
    public readonly roles = this.data.pipe(map((_) => _?.roles || []));
    public readonly contacts = this.data.pipe(map((_) => _?.contacts || []));
    public readonly filtered_contacts = combineLatest([
        this.contacts,
        this.role_filter,
    ]).pipe(
        map(([list, role]) =>
            list.filter((_) => !role || _.roles.includes(role))
        )
    );

    public readonly copyToClipboard = (id: string) => {
        const success = this._clipboard.copy(id);
        if (success) notifySuccess("User's email copied to clipboard.");
    };

    constructor(
        private _org: OrganisationService,
        private _dialog: MatDialog,
        private _clipboard: Clipboard
    ) {}

    public ngOnInit() {}

    public manageRoles() {
        const ref = this._dialog.open(RoleManagementModalComponent, {});
        ref.afterClosed().subscribe(() => this._change.next(Date.now()));
    }

    public editContact(contact?: EmergencyContact) {
        const ref = this._dialog.open(EmergencyContactModalComponent, {
            data: contact,
        });
        ref.afterClosed().subscribe(() => this._change.next(Date.now()));
    }

    public async removeContact(contact: EmergencyContact) {
        const result = await openConfirmModal(
            {
                title: 'Remove Emergency Contact',
                content: `Are you sure you want to remove ${contact.name} from the emergency contacts?`,
                icon: { content: 'delete' },
            },
            this._dialog
        );
        if (result.reason !== 'done') return;
        result.loading('Removing contact...');
        const data: any = await this.data.pipe(take(1)).toPromise();
        const new_contacts = (data?.contacts || []).filter(
            (_) => _.id !== contact.id
        );
        await updateMetadata(this._org.building.id, {
            name: 'emergency_contacts',
            description: 'Emergency Contacts',
            details: { roles: data.roles, contacts: new_contacts },
        }).toPromise();
        result.close();
        this._change.next(Date.now());
        notifySuccess('Successfully removed emergency contact.');
    }
}
