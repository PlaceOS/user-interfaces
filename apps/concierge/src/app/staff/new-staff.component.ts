import { Component, OnDestroy, OnInit } from '@angular/core';

import { StaffStateService } from './staff-state.service';

@Component({
    selector: '[app-new-staff]',
    template: `
        <app-topbar></app-topbar>
        <div class="flex flex-1 h-px">
            <app-sidebar></app-sidebar>
            <main class="flex flex-col flex-1 w-1/2 h-full">
                <staff-topbar class="w-full"></staff-topbar>
                <staff-listings class="w-full flex-1 h-0"></staff-listings>
                <mat-progress-bar
                    class="w-full"
                    *ngIf="loading | async"
                    mode="indeterminate"
                ></mat-progress-bar>
            </main>
        </div>
    `,
    styles: [
        `
            :host {
                display: flex;
                flex-direction: column;
                height: 100%;
                width: 100%;
                background-color: var(--b1);
            }
        `,
    ],
})
export class NewStaffComponent implements OnInit, OnDestroy {
    public readonly loading = this._state.loading;

    constructor(private _state: StaffStateService) {}

    public ngOnInit() {
        this._state.startPolling();
    }

    public ngOnDestroy() {
        this._state.stopPolling();
    }
}
