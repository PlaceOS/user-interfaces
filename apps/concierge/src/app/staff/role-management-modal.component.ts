import { Component, ViewChild } from '@angular/core';
import { MatDialogRef } from '@angular/material/dialog';
import { CustomTooltipComponent } from '@placeos/components';
import { OrganisationService } from '@placeos/organisation';
import { showMetadata, updateMetadata } from '@placeos/ts-client';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { filter, map, shareReplay, switchMap, take } from 'rxjs/operators';

@Component({
    selector: 'role-management-modal',
    template: `
        <header>
            <h2>Manage Roles</h2>
            <button btn icon matRipple mat-dialog-close>
                <app-icon>close</app-icon>
            </button>
        </header>
        <main
            class="overflow-y-auto min-w-[20rem] divide-y divide-base-200 max-h-[65vh]"
        >
            <button
                btn
                matRipple
                class="flex items-center justify-center space-x-2 w-[calc(100%-1rem)] m-2"
                customTooltip
                (click)="active = ''; role_name = ''"
                [content]="role_form"
            >
                <div class="truncate">New Role</div>
                <app-icon>add</app-icon>
            </button>
            <ng-container *ngFor="let role of roles | async">
                <div
                    class="flex items-center space-x-2 hover:bg-base-200:bg-base-300 p-2"
                >
                    <div class="flex-1 truncate">{{ role }}</div>
                    <button
                        btn
                        icon
                        matRipple
                        (click)="active = role; role_name = role"
                        customTooltip
                        [content]="role_form"
                    >
                        <app-icon>edit</app-icon>
                    </button>
                    <button btn icon matRipple (click)="removeRole(role)">
                        <app-icon>delete</app-icon>
                    </button>
                </div>
            </ng-container>
        </main>
        <ng-template #role_form>
            <div class="bg-base-100 p-4 rounded ">
                <mat-form-field appearance="outline">
                    <input
                        matInput
                        [(ngModel)]="role_name"
                        placeholder="Role name"
                    />
                </mat-form-field>
                <button btn matRipple class="w-full" (click)="updateRoles()">
                    Save Role
                </button>
            </div>
        </ng-template>
    `,
    styles: [``],
})
export class RoleManagementModalComponent {
    private _changes = new BehaviorSubject(0);

    public active: string;
    public role_name: string;
    public loading = false;
    public readonly data = combineLatest([
        this._org.active_building,
        this._changes,
    ]).pipe(
        filter(([bld]) => !!bld),
        switchMap(([bld]) => showMetadata(bld.id, 'emergency_contacts')),
        map(({ details }) => {
            const value = (details as any) || { roles: [], contacts: [] };
            if (!value.roles) value.roles = [];
            if (!value.contacts) value.contacts = [];
            return value;
        }),
        shareReplay(1)
    );
    public readonly roles = this.data.pipe(map((_) => _.roles));

    @ViewChild(CustomTooltipComponent) private _tooltip: CustomTooltipComponent;

    public async removeRole(role: string) {
        if (!role) return;
        this.loading = true;
        this._dialog_ref.disableClose = true;
        const data: any = await this.data.pipe(take(1)).toPromise();
        await updateMetadata(this._org.building.id, {
            name: 'emergency_contacts',
            description: 'Emergency Contacts',
            details: {
                roles: [...data.roles.filter((_) => _ !== role)]
                    .filter((_) => !!_)
                    .sort((a, b) => a.localeCompare(b)),
                contacts: data.contacts.map((_) => ({
                    ..._,
                    roles: _.roles.filter((r) => r !== role),
                })),
            },
        }).toPromise();
        this._changes.next(0);
        this.loading = false;
        this._dialog_ref.disableClose = false;
    }

    public async updateRoles() {
        if (!this.role_name) return;
        this.loading = true;
        this._tooltip.close();
        this._dialog_ref.disableClose = true;
        const data: any = await this.data.pipe(take(1)).toPromise();
        await updateMetadata(this._org.building.id, {
            name: 'emergency_contacts',
            description: 'Emergency Contacts',
            details: {
                roles: [
                    ...data.roles.filter((_) => _ !== this.active),
                    this.role_name,
                ]
                    .filter((_) => !!_)
                    .sort((a, b) => a.localeCompare(b)),
                contacts: data.contacts,
            },
        }).toPromise();
        this._changes.next(0);
        this.role_name = '';
        this.active = '';
        this.loading = false;
        this._dialog_ref.disableClose = false;
    }

    constructor(
        private _org: OrganisationService,
        private _dialog_ref: MatDialogRef<RoleManagementModalComponent>
    ) {}
}
