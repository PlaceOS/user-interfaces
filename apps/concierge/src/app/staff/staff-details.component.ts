import { Component, Input } from '@angular/core';
import { notifyError } from '@placeos/common';
import { StaffUser } from '@placeos/users';

import { StaffStateService } from './staff-state.service';

@Component({
    selector: 'staff-details',
    template: `
        <div
            class="w-full flex items-center px-4 py-2 bg-white border-b border-gray-200"
            *ngIf="user"
        >
            <a-user-avatar [user]="user"></a-user-avatar>
            <div class="flex flex-col flex-1">
                <div class="px-2">{{ user?.name }}</div>
                <div class="px-2 text-xs text-opacity-50">
                    {{ user?.email }}
                </div>
            </div>
            <div *ngIf="onsite" class="text-xs opacity-50 px-4">Onsite</div>
            <div class="flex items-center">
                <action-icon
                    [matTooltip]="onsite ? 'Checkout Staff' : 'Checkin Staff'"
                    [loading]="loading"
                    [content]="onsite ? 'event_busy' : 'event_available'"
                    (click)="onsite ? checkout() : checkin()"
                >
                </action-icon>
                <a
                    mat-icon-button
                    matTooltip="Email Staff"
                    [disabled]="!user?.email"
                    [href]="'mailto:' + user?.email"
                >
                    <app-icon>email</app-icon>
                </a>
                <a
                    mat-icon-button
                    matTooltip="Phone Staff"
                    [disabled]="!user?.phone"
                    [href]="'tel:' + user?.phone"
                >
                    <app-icon>call</app-icon>
                </a>
            </div>
        </div>
    `,
    styles: [
        `
            :host > div:hover {
                background-color: #f0f0f0 !important;
            }
        `,
    ],
})
export class StaffDetailsComponent {
    @Input() public user: StaffUser;
    @Input() public onsite: boolean;

    public loading: boolean;

    public readonly checkin = async () => {
        this.loading = true;
        await this._state
            .checkin(this.user)
            .catch((e) => notifyError('Error checking in Staff member'));
        this.loading = false;
    };
    public readonly checkout = async () => {
        this.loading = true;
        await this._state
            .checkout(this.user)
            .catch((e) => notifyError('Error checking out Staff member'));
        this.loading = false;
    };

    constructor(private _state: StaffStateService) {}
}
