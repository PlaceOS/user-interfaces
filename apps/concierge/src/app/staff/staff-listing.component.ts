import { Component, ElementRef, ViewChild } from '@angular/core';
import { BaseClass } from '@placeos/common';
import { map } from 'rxjs/operators';

import { StaffStateService } from './staff-state.service';

const CHARS = '#abcdefghijklmnopqrstuvwxyz'.split('');

@Component({
    selector: 'staff-listings',
    template: `
        <div class="w-full p-2 flex items-center justify-center">
            <div
                letter
                *ngFor="let group of groups"
                class="capitalize h-6 w-6 flex items-center justify-center text-xs cursor-pointer"
                [class.disabled]="(user_list | async)[group].length <= 0"
                [class.active]="group === active_group"
                (click)="scrollTo(group)"
            >
                {{ group }}
            </div>
        </div>
        <div
            class="flex-1 overflow-auto w-full relative bg-gray-100"
            style="height: 50%"
            #container
            (scroll)="onScroll($event)"
        >
            <ng-container *ngIf="user_count | async; else empty_state">
                <ng-container *ngFor="let group of groups">
                    <ng-container *ngIf="(user_list | async)[group].length">
                        <div
                            group
                            [id]="'letter-' + (group === '#' ? '0' : group)"
                            class="capitalize bg-gray-300 border-b text-sm font-medium sticky top-0 z-10"
                        >
                            {{ group }}
                        </div>
                        <staff-details
                            *ngFor="
                                let user of (user_list | async)[group];
                                let i = index
                            "
                            [id]="'letter-' + group + '-' + i"
                            [user]="user"
                            [onsite]="
                                (events | async)
                                    ? (events | async)[user.email]
                                    : false
                            "
                        ></staff-details>
                    </ng-container>
                </ng-container>
            </ng-container>
        </div>
        <mat-progress-bar
            *ngIf="loading | async"
            mode="indeterminate"
        ></mat-progress-bar>
        <ng-template #empty_state>
            <div
                class="absolute inset-0 flex flex-col items-center justify-center"
            >
                <p>No matching staff members</p>
            </div>
        </ng-template>
    `,
    styles: [
        `
            :host {
                display: flex;
                flex-direction: column;
                width: 100%;
                height: 50%;
            }

            [letter] {
                transition: font-size 200ms, color 200ms;
            }

            [group] {
                border-color: #ccc;
                padding: 0.5rem 1.65rem;
            }

            .disabled {
                opacity: 0.2;
                pointer-events: none;
            }

            .active {
                font-size: 1.25rem;
                opacity: 1;
                color: #d81b60;
            }
        `,
    ],
})
export class StaffListingComponent extends BaseClass {
    public active_group = '#';

    public readonly groups = CHARS;
    public readonly events = this._state.user_events;
    public readonly loading = this._state.loading;

    public readonly user_count = this._state.filtered_users.pipe(
        map((list) => list.length)
    );

    public readonly user_list = this._state.filtered_users.pipe(
        map((list) => {
            const user_map = {};
            for (const char of CHARS) {
                user_map[char] = (list || []).filter(
                    (user) =>
                        user.name.toLowerCase()[0].startsWith(char) ||
                        (char === '#' &&
                            !CHARS.includes(user.name.toLowerCase()[0]))
                );
            }
            this.timeout('scroll', () => this.onScroll({}), 30);
            return user_map;
        })
    );

    @ViewChild('container') private _el: ElementRef<HTMLDivElement>;

    constructor(private _state: StaffStateService) {
        super();
    }

    public onScroll(_) {
        const scroll_top = this._el.nativeElement.scrollTop;
        for (const group of CHARS) {
            const el: HTMLDivElement = document.querySelector(
                `#letter-${group === '#' ? '0' : group}`
            );
            if (el) {
                if (el.offsetTop - scroll_top > 0) {
                    break;
                }
                this.active_group = group;
            }
        }
    }

    public scrollTo(group: string) {
        const el = document.querySelector(`#letter-${group}-0`);
        if (el) {
            el.scrollIntoView({ behavior: 'smooth', block: 'center' });
            this.active_group = group;
        }
    }
}
