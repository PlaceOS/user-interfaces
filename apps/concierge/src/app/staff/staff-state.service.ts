import { Injectable } from '@angular/core';
import { searchStaff, StaffUser } from '@placeos/users';
import { BehaviorSubject, combineLatest } from 'rxjs';

import { Booking, checkinBooking, queryBookings, saveBooking } from '@placeos/bookings';
import { map, shareReplay, switchMap } from 'rxjs/operators';
import { endOfDay, startOfDay } from 'date-fns';
import { BaseClass, timePeriodsIntersect } from '@placeos/common';
import { HashMap } from '@placeos/ts-client/dist/esm/utilities/types';
import { OrganisationService } from '@placeos/organisation';

export interface StaffFilters {
    date?: number;
    zones?: string[];
    only_onsite?: boolean;
}

@Injectable({
    providedIn: 'root',
})
export class StaffStateService extends BaseClass {
    private _onsite: HashMap<boolean> = {};
    private _events: HashMap<Booking> = {};

    private _filters = new BehaviorSubject<StaffFilters>({});

    private _search = new BehaviorSubject<string>('');

    private _loading = new BehaviorSubject<boolean>(false);

    private _users = new BehaviorSubject<StaffUser[]>([]);

    public readonly loading = this._loading.asObservable();

    public readonly filters = this._filters.asObservable();

    public readonly users = this._filters.asObservable();

    public readonly filtered_users = combineLatest([
        this._search,
        this._users,
        this._filters,
    ]).pipe(
        map((details) => {
            const [filter, users, options] = details;
            return users.filter(
                (i) =>
                    (!filter ||
                        i.name.toLowerCase().includes(filter) ||
                        i.email.toLowerCase().includes(filter)) &&
                    (!options.only_onsite || this._onsite[i.email])
            );
        })
    );

    public readonly user_events = combineLatest([this._filters]).pipe(
        switchMap(async (_) => {
            this._loading.next(true);
            const bookings = await queryBookings({
                period_start: Math.floor(startOfDay(new Date()).valueOf() / 1000),
                period_end: Math.floor(endOfDay(new Date()).valueOf() / 1000),
                type: 'staff',
            }).toPromise();
            const checkin_map = {};
            const now = new Date().valueOf();
            for (const bkn of bookings) {
                if (
                    timePeriodsIntersect(
                        now,
                        now,
                        bkn.date,
                        bkn.date + bkn.duration * 60 * 1000
                    )
                ) {
                    checkin_map[bkn.asset_id] = bkn.checked_in;
                    this._events[bkn.asset_id] = bkn;
                }
            }
            this._onsite = checkin_map;
            this._loading.next(false);
            return checkin_map;
        }),
        shareReplay()
    );

    constructor(
        private _org: OrganisationService
    ) {
        super();
        this.loadUsers();
    }

    public setFilters(filters: StaffFilters) {
        this._filters.next({ ...this._filters.getValue(), ...filters });
    }

    public setSearchString(search: string) {
        this._search.next(search);
    }

    public startPolling(delay: number = 30 * 1000) {
        this.setFilters(this._filters.getValue());
        this.interval('poll', () => this.setFilters(this._filters.getValue()), delay);
    }

    public stopPolling() {
        this.clearInterval('poll');
    }

    public async checkin(user: StaffUser) {
        const result = await saveBooking({
            booking_start: Math.floor(
                new Date().valueOf() / 1000
            ),
            booking_end: Math.floor(endOfDay(new Date()).valueOf() / 1000),
            asset_id: user.email,
            title: 'Checked-in Onsite',
            description: this._org.building.display_name || this._org.building.name,
            zones: [this._org.building.id],
            booking_type: 'staff',
        } as any).toPromise();
        await checkinBooking(result.id, true);
        this._events[user.email] = result;
        this._onsite[user.email] = true;
    }

    public async checkout(user: StaffUser) {
        const event = this._events[user.email];
        if (event) {
            const result = await saveBooking({
                ...event.toJSON(),
                booking_end: Math.floor(new Date().valueOf() / 1000),
            } as any).toPromise();
            await checkinBooking(result.id, false).toPromise();
            this._events[user.email] = result;
            this._onsite[user.email] = false;
        }
    }

    private async loadUsers() {
        const user_list = await searchStaff('').toPromise();
        user_list.sort((a, b) => a.name.localeCompare(b.name));
        this._users.next(user_list);
    }
}
