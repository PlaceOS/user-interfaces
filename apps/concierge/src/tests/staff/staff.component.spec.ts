import { MatProgressBarModule } from '@angular/material/progress-bar';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { MockComponent } from 'ng-mocks';
import { BehaviorSubject } from 'rxjs';
import { StaffListingComponent } from '../../app/staff/staff-listing.component';
import { StaffStateService } from '../../app/staff/staff-state.service';
import { StaffTopbarComponent } from '../../app/staff/staff-topbar.component';
import { StaffComponent } from '../../app/staff/staff.component';
import { SidebarComponent } from '../../app/ui/sidebar.component';

describe('StaffComponent', () => {
    let spectator: Spectator<StaffComponent>;
    const createComponent = createComponentFactory({
        component: StaffComponent,
        providers: [
            {
                provide: StaffStateService,
                useValue: {
                    loading: new BehaviorSubject(false),
                    startPolling: jest.fn(),
                    stopPolling: jest.fn(),
                },
            },
        ],
        declarations: [
            MockComponent(SidebarComponent),
            MockComponent(StaffTopbarComponent),
            MockComponent(StaffListingComponent),
        ],
        imports: [MatProgressBarModule],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });
});
