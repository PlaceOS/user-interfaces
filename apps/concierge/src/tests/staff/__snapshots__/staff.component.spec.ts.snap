// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaffComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <sidebar />
  <main
    class="relative overflow-hidden flex-1 flex flex-col"
  >
    <staff-topbar
      class="w-full"
    />
    <staff-listings
      class="w-full flex-1 h-0"
    />
    
  </main>
</div>
`;
